LFS-273-275: Automatically flag answers/forms based on their values/levels of completion

IntegralProgrammer opened the pull request:
This PR replaces the **LFS-275** PR by incorporating the code revision commits from **LFS-273** up to commit `287c5e30e473a313c6c97463831b38042ccb5d83`.

The new features (**LFS-275**) added onto **LFS-273** are:

- When a form is instantiated, it starts out with `statusFlags = ["INCOMPLETE"]`
- When a form is saved, it is marked as `INCOMPLETE` if at least one descendant answer is marked as `INCOMPLETE`
- When a form is saved, it is marked as `INVALID` if at least one descendant answer is marked as `INVALID`
- When a form is saved, the `INCOMPLETE` flag is removed if no descendant answers possess this `INCOMPLETE` flag.
- When a form is saved, the `INVALID` flag is removed if no descendant answers possess this `INVALID` flag.

To test, build the `LFS-273-275_test` branch, and create the various _LFS-484_ test forms. Each form is configured so that every _Confirm Okay_ field has `minAnswers = 1`. Open `/bin/browser.html` and ensure that the `statusFlags` of the _root_ node for _each_ form are following the rules properly; `INCOMPLETE` for an initially blank _Confirm Okay_ answer, `INVALID,INCOMPLETE`, for a _Confirm Okay_ answer that was filled but then cleared.

IntegralProgrammer committed:
Replace property name string literals with property name string constants - to pass checkstyle tests

IntegralProgrammer committed:
AnswerStatusFlagEditor objects now hold a currentPath string variable which represents its location within the JCR repository. This feature will be used for moving up the JCR tree for form validation including forms with nested sections

IntegralProgrammer committed:
WIP

IntegralProgrammer committed:
WIP

IntegralProgrammer committed:
WIP

IntegralProgrammer committed:
Removed no longer necessary currentPath String variable

This has been replaced with the currentNodeBuilderPath ArrayList<NodeBuilder> variable

IntegralProgrammer committed:
Cleaned up commenting and documentation

IntegralProgrammer committed:
Fix issue with not properly setting statusFlags of a Form when an answer property is deleted

IntegralProgrammer committed:
Now properly initializes form with a status flag of INCOMPLETE

IntegralProgrammer committed:
Remove debugging log statements

IntegralProgrammer committed:
Automatically set statusFlags of each form to INCOMPLETE upon creation, even before SAVE button is clicked

sashaandjic reviewed (approved):
tested

IntegralProgrammer committed:
WIP: Do not include a child node in a parent node summary if said child node is an AnswerSection and this section has been disabled by the user

IntegralProgrammer committed:
WIP

IntegralProgrammer committed:
Refactor: Create getChildNodeWithQuestion(NodeBuilder nb, String uuid) method

IntegralProgrammer committed:
Add Javadoc commenting for getSectionNode(NodeBuilder nb) method

IntegralProgrammer committed:
Refactor: create evalSectionCondition(PropertyState propA, Property propB, String operator) method

IntegralProgrammer committed:
Remove unnecessary commenting

IntegralProgrammer committed:
Beginning support for comparing different data types in evalSectionCondition method

IntegralProgrammer committed:
WIP

IntegralProgrammer committed:
Implementation of evaluateConditionNodeRecursive

IntegralProgrammer committed:
Add support for comparing property values of type STRING, LONG, DOUBLE, DECIMAL, BOOLEAN

IntegralProgrammer committed:
Remove commented-out code

IntegralProgrammer committed:
Add support for <> comparator operator

Robustness fix for propB.isMultiple()

IntegralProgrammer committed:
Add support for 'is empty' and 'is not empty' operators. Will need further testing.

IntegralProgrammer committed:
Add support for '<' and '>' operators

IntegralProgrammer committed:
WIP: Date comparison works for only '=', '<>' operators

IntegralProgrammer committed:
Add support for '<' '>' operators when performing date comparison

IntegralProgrammer committed:
Remove debug logging statements

IntegralProgrammer committed:
Remove unnecessary commenting

IntegralProgrammer committed:
Satisfy checkstyle:ReturnCount

IntegralProgrammer committed:
Satisfy checkstyle:NestedIfDepth

IntegralProgrammer committed:
Satisfy checkstyle:MultipleStringLiterals

IntegralProgrammer committed:
Satisfy checkstyle:CyclomaticComplexity

IntegralProgrammer committed:
Add sanitizeNodeName method

OrdiNeu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -0,0 +1,604 @@
+        if (questionNode != null && PROP_VALUE.equals(after.getName())) {
+            Iterable<String> nodeAnswers = after.getValue(Type.STRINGS);
+            int numAnswers = iterableLength(nodeAnswers);
+            ArrayList<String> statusFlags = new ArrayList<String>();
```
If I'm reading this right, it seems like we're storing both "invalid" and "incomplete" status under the same multi-valued variable, and relying on the order to keep them distinct. What's the advantage of doing this over storing it in two variables/properties of the node? As it is we need to continually do string comparisons, whereas we could instead just be checking boolean values.

IntegralProgrammer replied:
The "invalid" and "incomplete" status flags are stored under the same multi-valued variable, but we are *not* relying on order to keep them distinct. Instead, consider this multi-valued string variable as an unordered set. The presence (regardless of position) of the string `INCOMPLETE` marks the node as incomplete, and similarly, the presence (regardless of position) of the string `INVALID` marks the node as invalid - see the `summarizeBuilder()` function for an example of this in action.

I'm not sure of the reason why a multi-valued string variable was chosen instead of multiple single-valued boolean variables, but if I remember correctly, Sergiu had specified this. Can you please confirm/explain, @sdumitriu?

OrdiNeu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -0,0 +1,604 @@
+                try {
+                    summarizeBuilders(this.currentNodeBuilderPath);
+                } catch (RepositoryException e) {
+                    LOGGER.warn("Could not run summarize()");
```
```suggestion
                    LOGGER.warn("Could not run summarizeBuilders()");
                    LOGGER.warn(e.toString());
```
It might also be worth printing the stack trace while we're at it.

IntegralProgrammer replied:
Done :heavy_check_mark:

OrdiNeu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -0,0 +1,604 @@
+    private static final String PROP_VALUE = "value";
+    private static final String PROP_QUESTION = "question";
+
+    private static final String STATUS_FLAGS = "statusFlags";
```
```suggestion
    private static final String STATUS_FLAGS = "lfs:statusFlags";
```

IntegralProgrammer replied:
I think it might be confusing to do this renaming as `lfs:statusFlags` might be mistaken for a data type of status flags as opposed to simply the node property holding the status flags.

OrdiNeu commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -568,6 +568,8 @@
   // Hardcode the resource supertype: each form is a resource.
   - sling:resourceSuperType (STRING) = "lfs/Resource" mandatory autocreated protected
 
+  - statusFlags (STRING) = "INCOMPLETE" mandatory autocreated
```
```suggestion
  - lfs:statusFlags (STRING) = "INCOMPLETE" mandatory autocreated
```

IntegralProgrammer replied:
I think it might be confusing to do this renaming as `lfs:statusFlags` might be mistaken for a data type of status flags as opposed to simply the node property holding the status flags.

IntegralProgrammer committed:
Update modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java

Print a stack trace of the error when summarizeBuilders() fails

Co-authored-by: OrdiNeu <OrdinaryNeumann@gmail.com>

sdumitriu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -50,37 +69,53 @@
      * @param resourceResolver a ResourceResolver object used to obtain answer constraints
      */
-    public AnswerCompletionStatusEditor(NodeBuilder nodeBuilder, ResourceResolver resourceResolver)
+    public AnswerCompletionStatusEditor(ArrayList<NodeBuilder> nodeBuilder, ResourceResolver resourceResolver)
```
Concrete types should not be used if the implementation details are not important. You should use just `List`.

sdumitriu replied:
```suggestion
    public AnswerCompletionStatusEditor(List<NodeBuilder> nodeBuilder, ResourceResolver resourceResolver)
```

sdumitriu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -50,37 +69,53 @@
+    // and the last item corresponding to the current node. By keeping this list, one is capable of
+    // moving up the tree and setting status flags of ancestor nodes based on the status flags of a
+    // descendant node.
+    private final ArrayList<NodeBuilder> currentNodeBuilderPath;
```
```suggestion
    private final List<NodeBuilder> currentNodeBuilderPath;
```

sdumitriu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -50,37 +69,53 @@
      * Simple constructor.
      *
-     * @param nodeBuilder the builder for the current node
+     * @param nodeBuilder an ArrayList of NodeBuilder objects starting from the root of the JCR tree
```
```suggestion
     * @param nodeBuilder a list of NodeBuilder objects starting from the root of the JCR tree
```

sdumitriu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -50,37 +69,53 @@
+        if (questionNode != null && PROP_VALUE.equals(after.getName())) {
             Iterable<String> nodeAnswers = after.getValue(Type.STRINGS);
             int numAnswers = iterableLength(nodeAnswers);
             ArrayList<String> statusFlags = new ArrayList<String>();
```
```suggestion
            List<String> statusFlags = new ArrayList<String>();
```

sdumitriu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -101,24 +136,39 @@ public void propertyChanged(PropertyState before, PropertyState after) throws Co
+                    summarizeBuilders(this.currentNodeBuilderPath);
+                } catch (RepositoryException e) {
+                    LOGGER.warn("Could not run summarizeBuilders()");
+                    LOGGER.warn(e.toString());
```
```suggestion
                    LOGGER.warn("Could not run summarizeBuilders(): {}", e.getMessage, e);
```

sdumitriu committed:
LFS-192 / LFS-275: Automatically flag forms upon save based on their level of completion
Use the more generic List instead of ArrayList in variable and argument declarations

sdumitriu committed:
LFS-192 / LFS-275: Automatically flag forms upon save based on their level of completion
Applied codestyle

sdumitriu committed:
LFS-192 / LFS-275: Automatically flag forms upon save based on their level of completion
Better logging usage

sdumitriu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -197,18 +273,323 @@ private int iterableLength(Iterable iterable)
+            final Node keyANode = sectionNodeParent.getNode(keyA);
+            final String keyANodeUUID = keyANode.getIdentifier();
+            // Get the node from the Form containing the answer to keyANode
+            final NodeBuilder conditionalFormNode = getChildNodeWithQuestion(prevNb, keyANodeUUID);
```
This may be `null`.

IntegralProgrammer replied:
Fixed by commit `b7b40d4` :heavy_check_mark:

sdumitriu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -197,18 +273,323 @@ private int iterableLength(Iterable iterable)
+                }
+            }
+            return downstreamResult;
+        } else if ("lfs:Conditional".equals(conditionNode.getProperty("jcr:primaryType").getString())) {
```
The implementation assumes that `operandA` is always a reference to an answer, and `operandB` is a fixed value. While our existing conditions are like this, it is not always guaranteed to be so. The reason why `operandA` is a reference to an answer node is that it explicitly mentions that it is a reference, [for example in the Radiotherapy questionnaire](https://github.com/ccmbioinfo/lfs/blob/458f573870852110f0284d73370f550505125486/questionnaires/clinical-data/src/main/resources/SLING-INF/content/Radiotherapy.json#L20). You should make a separate function `Property getOperandValue(Node operand ... and other needed params)` which works like this:

* does the node have an `isReference` property set to true?
  * if no, just return the value of the `value` property
  * if yes, find the answer for the question named in the `value property`
    * if it exists, return the value of its `value` property
    * if not, return null

sdumitriu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -197,18 +273,323 @@ private int iterableLength(Iterable iterable)
+        return false;
+    }
+
+    private boolean getSectionCondition(final NodeBuilder nb, final NodeBuilder prevNb)
```
`get` implies a getter, but this evaluates the condition and checks if it is satisfied or not. Thus a better name for it would be `isConditionSatisfied`.

IntegralProgrammer replied:
Implemented by commit `250be76` :heavy_check_mark:

sdumitriu commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/AnswerCompletionStatusEditor.java:
```diff
@@ -197,18 +273,323 @@ private int iterableLength(Iterable iterable)
+            }
+        }
+        return false;
+    }
```
All this code should be in a separate class.

IntegralProgrammer committed:
Rename getSectionCondition to isConditionSatisfied

IntegralProgrammer committed:
Handle case inside of evaluateConditionNodeRecursive() where getChildNodeWithQuestion() returns null to avoid NullPointerException

IntegralProgrammer committed:
WIP: Add getOperandValue() function

IntegralProgrammer committed:
WIP: Add getOperandValue() function

IntegralProgrammer committed:
WIP: Add getOperandValue() function

IntegralProgrammer committed:
WIP: Add getOperandValue() function

IntegralProgrammer committed:
WIP: Add getOperandValue() function

IntegralProgrammer committed:
Remove debug logging statements

IntegralProgrammer committed:
Re-factor: Move code from AnswerCompletionStatusEditor.java into ConditionalSectionUtils.java

IntegralProgrammer committed:
WIP: Implement back-end code for handling requireAll property of operand nodes (for multi answer questions)

IntegralProgrammer committed:
Fix bug where empty / non-empty answers were not being detected on the back-end

IntegralProgrammer committed:
Add support for < and > operators with BigDecimal data types

IntegralProgrammer committed:
Remove unnecessary commenting

IntegralProgrammer committed:
Remove debug logging statements

sashaandjic reviewed (approved):
tested all scenarios back and fourth in LFS-273-275_test branch

sdumitriu committed:
LFS-275: Automatically flag forms upon save based on their level of completion
Only compute the form status once per commit

sdumitriu committed:
LFS-275: Automatically flag forms upon save based on their level of completion
Removed unneeded check

sdumitriu committed:
[misc] Improved codestyle

sdumitriu committed:
LFS-275: Automatically flag forms upon save based on their level of completion
Empty forms and sections are considered incomplete

sdumitriu committed:
LFS-275: Automatically flag forms upon save based on their level of completion
Comparing values should always force the two operands to the right type

IntegralProgrammer commented:
I have found a bug introduced with the refactoring done by the latest _force-push_. To reproduce it:

- Build the `LFS-273-275_decimal_test` branch and start LFS
- Create a new **LFS-484 (Decimal) Test** form.
- Set **Must be equal to 22.5** to **22.5**
- Set **Confirm okay** to **ok**
- Click **Save** ... This succeeds with **HTTP 200**.

---

- Build the `LFS-273-275_refactor_decimal_test` branch and start LFS
- Create a new **LFS-484 (Decimal) Test** form.
- Set **Must be equal to 22.5** to **22.5**
- Set **Confirm okay** to **ok**
- Click **Save** ... This fails with **HTTP 500**.

sashaandjic commented:
![image](https://user-images.githubusercontent.com/23243244/91910852-ca75b400-ec7d-11ea-8492-8919b071cf41.png)

sdumitriu committed:
LFS-275: Automatically flag forms upon save based on their level of completion
Fixed NPE when comparing dates and decimals
Refactored the comparison methods to remove some code duplication

sdumitriu commented:
@IntegralProgrammer @sashaandjic This should be fixed now, please re-review

IntegralProgrammer commented:
I have found another bug. To reproduce it:

1. Build the `LFS-273-275_test` branch
2. Create a **LFS-484 (Decimal) Test** form
3. Set **Must be equal to 22.5** to **22.5**
4. Set **Confirm okay** to **ok**
5. Click **Save**
6. Clear the contents of the **Confirm okay** field
7. Click **Save**
8. Open `/bin/browser.html` and look under `/Forms/` in the JCR tree. The Form that we have just created/edited will be flagged as `INCOMPLETE`, even though it should be `INVALID, INCOMPLETE`. Looking further down the tree, you will be able to see that the `INVALID` flag is there but it is not being propagated upwards.

sashaandjic commented:
Confirming problem mentioned above.

IntegralProgrammer commented:
@sashaandjic I can reproduce the bug that you are describing. It's not exactly the odd - broken, even - working pattern that you are describing, but it is close. Specifically, for the **LFS-484 (Decimal) Test** form, the following is observed for the ability to properly propagate the `INVALID` flag upwards:

- Must be equal to 22.5 :x: 
- Must NOT be equal to 22.5 :heavy_check_mark: 
- Decimal Value A ... Decimal Value B :x: 
- Decimal Value C ... Decimal Value D :heavy_check_mark: 
- Decimal Value E ... Decimal Value F :x: 
- Decimal Value G ... Decimal Value H :x: 
- Must be empty :heavy_check_mark: 
- Must NOT be empty :heavy_check_mark: 

This bug does not seem to occur on the `LFS-273-275` branch prior to re-factoring. I believe that the cause of this issue is that the `==` , `<`, and `>` operators are always returning false and therefore the `INVALID` flag is conditionally blocked from propagating upwards. The negation of these operators always returns `True` and therefore allow the `INVALID` flag to propagate upwards (even when it shouldn't)

sashaandjic commented:
@IntegralProgrammer yes, that is correct. I was a bit lazy giving full details. It is not exactly odd - broken, even - working pattern.

sdumitriu committed:
LFS-275: Automatically flag forms upon save based on their level of completion
Fixed section evaluation for decimal values

IntegralProgrammer commented on modules/form-completion-status/src/main/java/ca/sickkids/ccm/lfs/formcompletionstatus/ConditionalSectionUtils.java:
```diff
@@ -0,0 +1,588 @@
+        if (propA == null || propB == null) {
+            return false;
+        }
+        Integer type = getPropertyObjectType(propA);
+        if (!operators.containsKey(type)) {
+            return false;
+        }
```
Should we also do this check for `propB`?

sdumitriu replied:
`propA` and `propB` are supposed to have compatible types, and while `propA` is mandatory, `propB` may be missing for `is [not] empty` checks. If `propB` has the wrong type, the questionnaire is malformed, and we shouldn't get into that state in the first place, and then we'll still get a proper comparison, even if with the wrong value.

IntegralProgrammer reviewed (commented):
@sdumitriu I have rebased the `LFS-273-275_test` branch based on this branch and all tests pass. Please have a look at my comment on adding an additional check for `propB`, but other than that, this seems good and I will approve. :heavy_check_mark:

sashaandjic reviewed (approved):
re-tested
